plugins/Oscilloscope/OscilloscopeGraph.cpp

```diff
@@ -0,0 +1,168 @@
+	p.drawLine(0, height() / 2 * (1 + amp), width(), height() / 2 * (1 + amp));
+
+	const bool hq = framesPerPixel > 1;
+	const int xoffset = !hq * framesPerPixel;
+	const float xscale = 1.f / (windowSizeFrames - framesPerPixel);
+	const bool stereo = m_controls->m_stereoModel.value();
+
+	auto drawWaveform = [&](QColor& color, auto&& getChannel)
+	{
+		p.setCompositionMode(QPainter::CompositionMode_Plus);
+		p.setPen(color);
+		for (int f = 0; f < windowSizeFrames - framesPerPixel; f += framesPerPixel)
+		{
+			const int currentIndex = (windowStartIndex + f) % m_ringBuffer.size();
+			const int nextIndex = (windowStartIndex + f + framesPerPixel) % m_ringBuffer.size();
+
+			float maxValue = getChannel(m_ringBuffer[hq ? currentIndex : nextIndex]);
+			float minValue = getChannel(m_ringBuffer[currentIndex]);
+
+			for (int i = currentIndex + 1; hq && i <= nextIndex; ++i)
+			{
+				maxValue = std::max(maxValue, getChannel(m_ringBuffer[i]));
+				minValue = std::min(minValue, getChannel(m_ringBuffer[i]));
+			}
+
+			p.drawLine(
+				width() * f * xscale,
+				height() * (1 - minValue * amp) / 2,
+				width() * (f + xoffset) * xscale,
+				height() * (1 - maxValue * amp) / 2
+			);
+		}
+	};
+
+	if (!stereo)
+	{
+		drawWaveform(m_monoColor, [](const SampleFrame& f){ return f.average(); });
+	}
+	else
+	{
+		drawWaveform(m_leftColor, [](const SampleFrame& f){ return f.left(); });
+		drawWaveform(m_rightColor, [](const SampleFrame& f){ return f.right(); });
+	}
+}
+
+void OscilloscopeGraph::wheelEvent(QWheelEvent* we)
+{
+	float windowSizeMilliseconds = m_controls->m_lengthModel.value();
+	float phase = m_controls->m_phaseModel.value();
+	float mouseOffset = (1.0f - we->position().x() / width()) * (windowSizeMilliseconds / m_controls->m_lengthModel.maxValue());
+	float zoomAmount = std::clamp(std::exp2(-we->angleDelta().y() / 240.0f), m_controls->m_lengthModel.minValue() / m_controls->m_lengthModel.value(), m_controls->m_lengthModel.maxValue() / m_controls->m_lengthModel.value());
+
+	if ((zoomAmount > 1.0f && windowSizeMilliseconds >= m_controls->m_lengthModel.maxValue()) || (zoomAmount < 1.0f && windowSizeMilliseconds <= m_controls->m_lengthModel.minValue())) { return; }
+
+	float newWindowSizeMilliseconds = windowSizeMilliseconds * zoomAmount;
+	float newPhase = phase - mouseOffset * (1.0f - zoomAmount);
+	m_controls->m_lengthModel.setValue(newWindowSizeMilliseconds);
+	// Clamp to prevent the user from accidentally bringing the write position discontinuity into view
+	m_controls->m_phaseModel.setValue(std::clamp(newPhase, newWindowSizeMilliseconds / m_controls->m_lengthModel.maxValue(), 1.0f));
+}
+
+void OscilloscopeGraph::mousePressEvent(QMouseEvent* me)
+{
+	m_mousePos = me->x();
+}
+
+void OscilloscopeGraph::mouseMoveEvent(QMouseEvent* me)
+{
+	float phase = m_controls->m_phaseModel.value();
+	float newPhase = phase + 1.0f * (m_mousePos - me->x()) / width() * (m_controls->m_lengthModel.value() / m_controls->m_lengthModel.maxValue());
```

Comment:
- What is the purpose of the multiplication by `1.0f` on line 162?
- (Non-blocking) `newPhase` is never mutated and should be declared `const`
- (Non-blocking) `phase` is used exactly once and can be inlined (it's clear that it's the phase even without this variable binding)

```suggestion
	float newPhase = m_controls->m_phaseModel.value() + (m_mousePos - me->x()) / width() * (m_controls->m_lengthModel.value() / m_controls->m_lengthModel.maxValue());
```

Reply:
I'm guessing the `1.0f` was to prevent integer division from happening when calculating the ratio of the mouse offset pixels to the width of the oscilloscope graph. I will test it locally to see if it works fine without it.